File: src/middleware/logger-middleware.ts
import { Middleware } from "../types";
import { entries } from "../utils/entries";

/**
 * A middleware that logs every action that is dispatched, and the new state
 * after the action has been dispatched.
 */
export const loggerMiddleware: Middleware = (producer) => (done) => (action) => {
	const params: string[] = [];

	for (const [, value] of entries(action.arguments)) {
		params.push(stringify(value));
	}

	print(`[loggerMiddleware]: Dispatching ${action.type}(${params.join(", ")})`);

	const newState = done(action);

	print("[loggerMiddleware]: New state:", producer.getState());

	return newState;
};

function stringify(value: unknown): string {
	if (typeIs(value, "string")) {
		return `"${value}"`;
	} else if (typeIs(value, "table")) {
		return `{ ${[...entries(value)].map(([key, value]) => `${key}: ${tostring(value)}`).join(", ")} }`;
	}

	return tostring(value);
}

File: src/middleware/loggerMiddleware.lua
local types = require(script.Parent.Parent.types)
local logger = require(script.Parent.Parent.utils.logger)

local loggerMiddleware: types.Middleware = function(producer)
	logger.info("Initial state:", producer:getState())

	producer:subscribe(function(state)
		logger.info("State changed:", state)
	end)

	return function(dispatch, name)
		return function(...)
			local output = { `Dispatching {name}\nArguments:` }

			for index = 1, select("#", ...) do
				table.insert(output, `\n\t{index}.`)
				table.insert(output, select(index, ...))
			end

			logger.info(table.unpack(output))

			return dispatch(...)
		end
	end
end

return loggerMiddleware
